scripts/bert/run_pretraining.py

```diff
@@ -31,81 +31,170 @@
+                         'Effective only if --raw is set.')
+parser.add_argument('--sp_nbest', type=int, default=0,
+                    help='Number of best candidates for sampling subwords with sentencepiece. '
+                         'Effective only if --raw is set.')
+parser.add_argument('--sp_alpha', type=float, default=1.0,
+                    help='Inverse temperature for probability rescaling for sentencepiece '
+                         'sampling. Effective only if --raw is set.')
+parser.add_argument('--num_data_workers', type=int, default=8,
+                    help='Number of workers to pre-process data. '
+                         'Effective only if --raw is set.')
+parser.add_argument('--eval_use_npz', action='store_true',
+                    help='Set to True if --data_eval provides npz files instead of raw text files')
+parser.add_argument('--backend', type=str, default='device',
+                    choices=['horovod', 'dist_sync_device', 'device'],
+                    help='Communication backend.')
 
-os.environ['MXNET_KVSTORE_USETREE'] = '1'
-os.environ['MXNET_GPU_MEM_POOL_TYPE'] = 'Round'
+args = parser.parse_args()
 
 # logging
 level = logging.DEBUG if args.verbose else logging.INFO
 logging.getLogger().setLevel(level)
 logging.info(args)
+os.environ['MXNET_GPU_MEM_POOL_TYPE'] = 'Round'
 
-class ParallelBERT(nlp.utils.Parallelizable):
+class DataParallelBERT(nlp.utils.Parallelizable):
     """Data parallel BERT model.
 
     Parameters
     ----------
     model : Block
         The BERT model.
     """
-    def __init__(self, model, mlm_loss, nsp_loss, vocab_size, rescale_factor, trainer=None):
+    def __init__(self, model, trainer):
         self._model = model
-        self._mlm_loss = mlm_loss
-        self._nsp_loss = nsp_loss
-        self._vocab_size = vocab_size
-        self._rescale_factor = rescale_factor
         self._trainer = trainer
 
     def forward_backward(self, x):
         """forward backward implementation"""
+        (input_id, masked_id, masked_position, masked_weight, \
+         next_sentence_label, segment_id, valid_length) = x
+
+        valid_length = valid_length.astype(args.dtype, copy=False)
         with mx.autograd.record():
-            (ls, next_sentence_label, classified, masked_id, decoded, \
-             masked_weight, ls1, ls2, valid_length) = forward(x, self._model, self._mlm_loss,
-                                                              self._nsp_loss, self._vocab_size,
-                                                              args.dtype)
-            ls = ls / self._rescale_factor
+            out = self._model(input_id, masked_id, masked_position, masked_weight,
+                              next_sentence_label, segment_id, valid_length)
+            classified, decoded, ls1, ls2 = out
+            ls = ls1 + ls2
+            ls = ls / args.accumulate
         if args.dtype == 'float16':
             self._trainer.backward(ls)
```

Comment:
This logic is a little bit strange here. Is it true that when args.dtype == 'float16', _trainer is always expected? If so, maybe write `if _trainer is not None:` ?